src/main/java/org/apache/datasketches/kll/KllMemoryValidate.java

```diff
@@ -45,157 +32,162 @@
+  long n;        //8 bytes (if present), used multiple places
+  int minK;      //2 bytes (if present), used multiple places
+  int numLevels; //1 byte  (if present), used by KllPreambleUtil
+  //skip unused byte
+  int[] levelsArr; //starts at byte 20, adjusted to include top index here, used multiple places
+
+  // derived.
+  int sketchBytes = 0; //used by KllPreambleUtil
+  private int typeBytes = 0; //always 0 for generic
+
   KllMemoryValidate(final Memory srcMem, final SketchType sketchType) {
+    this(srcMem, sketchType, null);
+  }
+
+  KllMemoryValidate(final Memory srcMem, final SketchType sketchType, final ArrayOfItemsSerDe<?> serDe) {
+    final long memCapBytes = srcMem.getCapacity();
+    if (memCapBytes < 8) { throw new SketchesArgumentException(MEMORY_TOO_SMALL + memCapBytes); }
+    this.srcMem = srcMem;
+    this.sketchType = sketchType;
+    this.serDe = serDe;
     preInts = getMemoryPreInts(srcMem);
     serVer = getMemorySerVer(srcMem);
+    sketchStructure = SketchStructure.getSketchStructure(preInts, serVer);
     familyID = getMemoryFamilyID(srcMem);
-    if (familyID != Family.KLL.getID()) { memoryValidateThrow(SRC_NOT_KLL, familyID); }
+    if (familyID != Family.KLL.getID()) { throw new SketchesArgumentException(SRC_NOT_KLL + familyID); }
     flags = getMemoryFlags(srcMem);
     k = getMemoryK(srcMem);
     m = getMemoryM(srcMem);
     KllHelper.checkM(m);
     KllHelper.checkK(k, m);
-    
-    empty = getMemoryEmptyFlag(srcMem);
-    level0Sorted  = getMemoryLevelZeroSortedFlag(srcMem);
-    singleItem = getMemorySingleItemFlag(srcMem);
-    
-    serialVersionUpdatable = serVer == SERIAL_VERSION_UPDATABLE;
-    typeBytes = (sketchType == DOUBLES_SKETCH) ? Double.BYTES : Float.BYTES;
-
-    if (serialVersionUpdatable) { updatableMemFormatValidate((WritableMemory) srcMem); }
-    else { compactMemoryValidate(srcMem); }
+    //flags
+    emptyFlag = getMemoryEmptyFlag(srcMem);
+    level0SortedFlag  = getMemoryLevelZeroSortedFlag(srcMem);
+    if (sketchType == DOUBLES_SKETCH) { typeBytes = Double.BYTES; }
+    else if (sketchType == FLOATS_SKETCH) { typeBytes = Float.BYTES; }
+    else { typeBytes = 0; }
+    validate();
   }
 
-  private void compactMemoryValidate(final Memory srcMem) { //FOR HEAPIFY.  NOT UPDATABLE
-    if (empty && singleItem) { memoryValidateThrow(EMPTYBIT_AND_SINGLEBIT, flags); }
-    final int sw = (empty ? 1 : 0) | (singleItem ? 4 : 0);
+  private void validate() {
 
-    switch (sw) {
-      case 0: { //FULL_COMPACT
-        if (preInts != PREAMBLE_INTS_FULL) { memoryValidateThrow(INVALID_PREINTS, preInts); }
-        if (serVer != SERIAL_VERSION_EMPTY_FULL) { memoryValidateThrow(EMPTYBIT_AND_SER_VER, serVer); }
+    switch (sketchStructure) {
+      case COMPACT_FULL: {
+        if (emptyFlag) { throw new SketchesArgumentException(EMPTY_FLAG_AND_COMPACT_FULL); }
         n = getMemoryN(srcMem);
+        //if (n <= 1) { memoryValidateThrow(N_AND_COMPACT_FULL); }
```

Comment:
do we need ti keep this commented-out line?

Reply:
Ah, this was a result of a discussion we had WRT an "improperly" compacted sketch.  This is where a sketch has only one value, but was not compacted using the Single Item serialization layout.  I think we concluded that this could happen with a very old sketch built prior to the introduction of the Single Item layout.  

I added a comment to that effect so that we can remember why we can't use this test.  I also commented out
the associated error message string at the bottom of the class.